tests/utils.ts

```diff
@@ -156,3 +221,65 @@ export async function sendWsMessage(
     };
   });
 }
+
+const MAX_RESPONSE_TIMEOUT = 1_500;
+
+export async function subscriptionRequest(
+  url: string,
+  request: JsonRpcRequest,
+  messageHandler: MessageSubscriptionHandler
+): Promise<{ status: any, subscription?: { id: string, close: () => Promise<void> } }> {
+  let resolved: boolean = false;
+  const { id: requestId } = request;
+  const connection = await JsonRpcClient.connect(url);
+
+  const closeSubscription = async (id: JsonRpcId, connection: JsonRpcClient): Promise<JsonRpcResponse> => {
+    const requestId = uuidv4();
+    const request = createJsonRpcRequest(requestId, 'subscription.close', { id });
+    return await connection.request(request);
+  }
+
+  return new Promise<{ status: any, subscription?: { id: string, close: () => Promise<void> } }>((resolve, reject) => {
+    const { close: subscriptionClose } = connection.subscribe(request, (response) => {
+      const { result, error } = response;
+
+      // this is an error specific to the `JsonRpcRequest` requesting the subscription
+      if (error) {
+        reject(error);
+        return;
+      }
+
+      // at this point the reply should be DwnRpcResponse
+      const { status, record, subscription } = result.reply;
+      if (record) {
+        messageHandler(record);
+        return;
+      }
+      if (subscription) {
+        resolved = true;
+        resolve({
+          status,
+          subscription: {
+            ...subscription,
+            close: async (): Promise<void> => {
+              subscriptionClose();
+              const closeResponse = await closeSubscription(requestId, connection);
+              if (closeResponse.error?.message !== undefined) {
+                throw new Error(`unable to close subscription: ${closeResponse.error.message}`);
+              }
+            }
+          }
+        })
+      } else {
+        resolve({ status });
+      }
+    });
+
+    setTimeout(() => {
+      if (resolved) {
+        return;
+      };
```

Comment:
I think no need for this check and the `resolved` variable, once resolved, calling `reject` has no effect.

Reply:
Yup! This was copy-pasta, I had it doing a hacky Promise.race before.

Reply:
Done